Authenticators/Elkarte/elkarteauth.py

```diff
@@ -0,0 +1,887 @@
+            """
+
+            FALL_THROUGH = ""
+            
+            debug('idToTexture for %d', id)
+            if id < cfg.user.id_offset or not cfg.user.avatar_enable:
+                debug('idToTexture %d -> fall through', id)
+                return FALL_THROUGH
+            
+            # Otherwise get the users texture from elkarte
+            bbid = id - cfg.user.id_offset
+            try:
+                sql = 'SELECT avatar FROM %smembers WHERE id_member = %%s' % cfg.database.prefix
+                cur = threadDB.execute(sql, [bbid])
+            except threadDbException:
+                return FALL_THROUGH
+            res = cur.fetchone()
+            cur.close()
+            if not res:
+                debug('idToTexture %d -> user unknown, fall through', id)
+                return FALL_THROUGH
+            avatar = res[0]
+            
+            if not avatar:
+                # Either the user has none or it is in the attachments, check there
+                try:
+                    sql = '''SELECT id_attach, file_hash, filename, attachment_type FROM %sattachments WHERE approved = true AND
+                        (attachment_type = 0 OR attachment_type = 1) AND id_member = %%s''' % cfg.database.prefix
+                    cur = threadDB.execute(sql, [bbid])
+                except threadDbException:
+                    return FALL_THROUGH
+                
+                res = cur.fetchone()
+                cur.close()
+                if not res:
+                    # No uploaded avatar found, seems like the user didn't set one
+                    debug('idToTexture %d -> no texture available for this user, fall through', id)
+                    return FALL_THROUGH
+                
+                fid, fhash, filename, fattachtype = res
+                if cfg.forum.path.startswith('file://'):
+                    # We are supposed to load this from the local fs
+                    avatar_file = cfg.forum.path + 'attachments/%d_%s' % (fid, fhash)
+                elif fattachtype == 0: 
+                    avatar_file = cfg.forum.path + 'index.php?action=dlattach;attach=%d;type=avatar' % fid
+                elif fattachtype == 1:
+                    avatar_file = cfg.forum.path + 'avatars/' + filename 
+            elif "://" in avatar:
+                # ...or it is a external link
+                avatar_file = avatar
+            else:
+                warning("avatar with an unexpected value, fall through")
+                return FALL_THROUGH
+                
+            if avatar_file in self.texture_cache:
+                return self.texture_cache[avatar_file]
+            
+            try:
+                handle = urllib2.urlopen(urllib2.Request(avatar_file, headers={'Accept': 'text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8', 'User-Agent': 'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US; rv:1.9.0.7) Gecko/2009021910 Firefox/3.0.7'}))
```

Comment:
Wow. What a useless feature. From what I could find it's part of their "Browser Integrity Check". While changing the headers might work for now judging from [this description](https://support.cloudflare.com/hc/en-us/articles/200170086-What-does-the-Browser-Integrity-Check-do-) as well as [this](http://cstress.net/blog/cloudflares-browser-integrity-check-replicate/) that might not last and you might get challenged anyway as long as the feature is enabled.

It seems you can manually [trust certain IPs](https://support.cloudflare.com/hc/en-us/articles/200171346-How-do-I-block-blacklist-or-trust-whitelist-an-IP-address-manually-) to exclude them from cloudflares threat management which is probably more of a "the way they intended it to be used" solution. I assume you can configure that on their free accounts too.

If we stick with the "we allow spoofing" I would probably want that to be part of the configuration with the default being to behave nicely and report a custom user-agent (e.g. "elkarteauth" here) etc. and the alternative maybe being in there but commented out by default telling the user why he might need to mess with it. Having benign scripts with very specific capabilities pretend to be a full blown browser by default feels like asking for trouble.

Thoughts?